File: benchmarks/dummy_backend.py
#!/usr/bin/env python3
"""Minimal fixed-response HTTP server, for benchmarking the gateway against
a uniform pool of identical backends (instead of two different real services).

Usage: python3 dummy_backend.py <port> [service_name] [body_bytes]

The optional service_name is echoed back in the response body so you can
confirm path-based routing is sending requests to the right service, and
which replica within that service answered.

body_bytes sizes the response payload, defaulting to 65536 — the gateway's
HttpObjectAggregator limit (64 * 1024) on both the request and the backend
response side. Asking for more than that makes the gateway reject the
response rather than forward it.

The payload is random alphanumeric data, generated once at startup: doing it
per request would make Python's RNG the bottleneck instead of the gateway.
"""
import sys
import json
import random
import string
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 9001
SERVICE = sys.argv[2] if len(sys.argv) > 2 else "dummy"
BODY_BYTES = int(sys.argv[3]) if len(sys.argv) > 3 else 64 * 1024


def build_body(target_bytes):
    """Return a JSON body of exactly target_bytes, padded with random data."""
    envelope = {"service": SERVICE, "port": PORT, "payload": ""}
    overhead = len(json.dumps(envelope).encode("utf-8"))
    # Alphanumerics need no JSON escaping, so one character is one byte.
    pad = max(0, target_bytes - overhead)
    envelope["payload"] = "".join(
        random.choices(string.ascii_letters + string.digits, k=pad)
    )
    return json.dumps(envelope).encode("utf-8")


BODY = build_body(BODY_BYTES)


class Handler(BaseHTTPRequestHandler):
    # BaseHTTPRequestHandler defaults to HTTP/1.0, which closes the connection
    # after every response. Keep-alive is required for the gateway's connection
    # pool to have anything to reuse, so opt into HTTP/1.1.
    protocol_version = "HTTP/1.1"

    def do_GET(self):
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(BODY)))
        self.end_headers()
        self.wfile.write(BODY)

    def do_POST(self):
        self.do_GET()

    def log_message(self, format, *args):
        pass  # silence per-request logging


if __name__ == "__main__":
    server = ThreadingHTTPServer(("0.0.0.0", PORT), Handler)
    print(f"dummy backend '{SERVICE}' listening on :{PORT} ({len(BODY)} byte body)")
    server.serve_forever()

File: benchmarks/dummy_backend.go
// Minimal fixed-response HTTP server, for benchmarking the gateway against
// a uniform pool of identical backends.
//
// Usage: dummy_backend <port> [service_name] [body_bytes]
//
// Replaces the old Python (http.server) version: Go's net/http has no
// small default listen backlog (Python's socketserver.TCPServer defaults
// to 5, which resets connections under high concurrency regardless of
// CPU headroom) and no GIL serializing request handling within a process.
//
// The payload is random alphanumeric data, generated once at startup: doing
// it per request would make the RNG the bottleneck instead of the gateway.
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"net"
	"net/http"
	"os"
	"strconv"
)

const letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

func randomString(n int) string {
	b := make([]byte, n)
	for i := range b {
		b[i] = letters[rand.Intn(len(letters))]
	}
	return string(b)
}

func buildBody(service string, port int, targetBytes int) []byte {
	envelope := map[string]interface{}{"service": service, "port": port, "payload": ""}
	overhead, _ := json.Marshal(envelope)
	pad := targetBytes - len(overhead)
	if pad < 0 {
		pad = 0
	}
	envelope["payload"] = randomString(pad)
	body, _ := json.Marshal(envelope)
	return body
}

func main() {
	port := 9001
	service := "dummy"
	bodyBytes := 64 * 1024

	if len(os.Args) > 1 {
		port, _ = strconv.Atoi(os.Args[1])
	}
	if len(os.Args) > 2 {
		service = os.Args[2]
	}
	if len(os.Args) > 3 {
		bodyBytes, _ = strconv.Atoi(os.Args[3])
	}

	body := buildBody(service, port, bodyBytes)

	handler := func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("Content-Length", strconv.Itoa(len(body)))
		w.WriteHeader(http.StatusOK)
		w.Write(body)
	}

	addr := fmt.Sprintf("0.0.0.0:%d", port)
	// Explicit listener + large backlog (Go's default is already the OS max
	// via net.ListenConfig, but spell it out since the small Python backlog
	// was the whole reason for this rewrite).
	lc := net.ListenConfig{}
	ln, err := lc.Listen(nil, "tcp", addr)
	if err != nil {
		log.Fatalf("failed to listen on %s: %v", addr, err)
	}

	fmt.Printf("dummy backend '%s' listening on :%d (%d byte body)\n", service, port, len(body))
	server := &http.Server{Handler: http.HandlerFunc(handler)}
	log.Fatal(server.Serve(ln))
}
